src/main/java/emissary/parser/NIOSessionParser.java

```diff
@@ -11,22 +11,38 @@
 import org.slf4j.LoggerFactory;
 
 /**
- * Provide a basic NIO based session parser.
+ * Provide a basic NIO-based session parser that reads data in chunks from the underlying channel. A chunk might have
+ * zero or more complete sessions within it. The chunk buffer will begin at minChunkSize, but grow as large as
+ * maxChunkSize in order to accomodate a complete session. Sessions larger than maxChunkSize will lead to
+ * ParserExceptions
  */
 public abstract class NIOSessionParser extends SessionParser {
     // Logger
     private final static Logger logger = LoggerFactory.getLogger(NIOSessionParser.class);
 
-    // the input channel we will read from
+    protected static final int MIN_CHUNK_SIZE_DEFAULT = 2 * 1024 * 1024; // 2Mb
+    protected static final int MAX_CHUNK_SIZE_DEFAULT = 40 * 1024 * 1024; // 40Mb
+
+    /** The data source for this parser */
     protected SeekableByteChannel channel;
 
-    // Track where we are in the overall file and in the current chunk
+    /** The start position of the current chunk relative to the data source */
     protected int chunkStart = 0;
+
+    /** The current chunk buffer */
     protected byte[] data = null;
 
-    // Max chunk to read at a time, we will never be able to read a file with a session larger than this.
-    protected static final int MAP_MAX_DEFAULT = 40 * 1024 * 1024; // 40Mb
-    protected int chunkSize = MAP_MAX_DEFAULT;
+    /** The current write position for the current chunk buffer */
+    protected int writeOffset = 0;
+
+    /** Min chunk buffer size. */
+    protected int minChunkSize = MIN_CHUNK_SIZE_DEFAULT;
+
+    /** Max chunk to read at a time, we will never be able to read a file with a session larger than this. */
+    protected int maxChunkSize = MAX_CHUNK_SIZE_DEFAULT;
+
+    /** When we grow the chunk buffer to accomodate additional data, we will grow the buffer by this increment */
+    protected int chunkAllocationIncrement = (10 * 1024 * 1024) + 100;
```

Comment:
Why 100MB + 100?

Reply:
This is sort of arbitrary, but I did not want to make it an even multiple of the default maxChunkSize, as the comments suggest - I'm shooting for something along the lines of ByteArrayOutputStream's `grow()` implementation eventually, but that will be another PR